File: CommandCPP/LightCommand.cpp
/* Copyright (c) 2022 RodKingroo */
#include "LightCommand.h"

void setUndo(shared_ptr<Light> light, int prevBright)
{
    switch (prevBright)
    {
        case Light::OFF:
            light->off();
        case Light::STATE_ONE:
            light->state_1();
            break;
        case Light::STATE_TWO:
            light->state_2();
            break;
        case Light::STATE_THREE:
            light->state_3();
            break;
        case Light::STATE_FOUR:
            light->state_4();
            break;
        case Light::STATE_FIVE:
            light->state_5();
            break;

    }
}


/* Light On State One Command */
LightOnStateOneCommand::LightOnStateOneCommand(shared_ptr<Light> light)
{
    this->light = light;
}

void LightOnStateOneCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_1();
}

void LightOnStateOneCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}



/* Light on State Two Command */
LightOnStateTwoCommand::LightOnStateTwoCommand(shared_ptr<Light> light)
{
    this->light = light;

}

void LightOnStateTwoCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_2();

}

void LightOnStateTwoCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}


/* Light on State Three Command */
LightOnStateThreeCommand::LightOnStateThreeCommand(shared_ptr<Light> light)
{
    this->light = light;
}

void LightOnStateThreeCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_3();

}

void LightOnStateThreeCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}


/* Light on State Three Command */
LightOnStateFourCommand::LightOnStateFourCommand(shared_ptr<Light> light)
{
    this->light = light;
}

void LightOnStateFourCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_4();
}

void LightOnStateFourCommand::undo() /* override */
{
    light->off();  
    setUndo(light, prevBright);
}


/* Light on State Five Command */
LightOnStateFiveCommand::LightOnStateFiveCommand(shared_ptr<Light> light)
{
    this->light = light;
}

void LightOnStateFiveCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_5();
}

void LightOnStateFiveCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}


/* Light on Off Command */
LightOffCommand::LightOffCommand(shared_ptr<Light> light)
{
    this->light = light;
}

void LightOffCommand::execute() /* override */
{
    prevBright = light->getState();
    light->off();
}

void LightOffCommand::undo() /* override */
{
    light->on();
    setUndo(light, prevBright);
}

File: CommandPython/LightCommand.py
# Copyright (c) 2022 RodKingroo

from Light import *
from Command import *

def setUndo(light, prevBright):
    if prevBright == light.STAGE_ONE:
        light.stage_1()
    elif prevBright == light.STAGE_TWO:
        light.stage_2()
    elif prevBright == light.STAGE_THREE:
        light.stage_3()
    elif prevBright == light.STAGE_FOUR:
        light.stage_4()
    elif prevBright == light.STAGE_FIVE:
        light.stage_5()
        
    
# Light on Stage One command
class LightOnStageOneCommand(Command):
    def __init__(self, light):
        if isinstance(light, Light):
            self.light = light
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.light.getState()
        self.light.on()
        self.light.stage_1()
        
    def undo(self):
        self.light.off()
        self.light = setUndo(self.light, self.prevBright)
        

# Light on Stage Two command
class LightOnStageTwoCommand(Command):
    def __init__(self, light):
        if isinstance(light, Light):
            self.light = light
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.light.getState()
        self.light.on()
        self.light.stage_2()
        
    def undo(self):
        self.light.off()
        self.light = setUndo(self.light, self.prevBright)
        
# Light on Stage Three command
class LightOnStageThreeCommand(Command):
    def __init__(self, light):
        if isinstance(light, Light):
            self.light = light
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.light.getState()
        self.light.on()
        self.light.stage_3()
        
    def undo(self):
        self.light.off()
        self.light = setUndo(self.light, self.prevBright)
        
# Light on Stage Four command
class LightOnStageFourCommand(Command):
    def __init__(self, light):
        if isinstance(light, Light):
            self.light = light
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.light.getState()
        self.light.on()
        self.light.stage_4()
        
    def undo(self):
        self.light.off()
        self.light = setUndo(self.light, self.prevBright)
        
# Light on Stage Five command
class LightOnStageFiveCommand(Command):
    def __init__(self, light):
        if isinstance(light, Light):
            self.light = light
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.light.getState()
        self.light.on()
        self.light.stage_5()
        
    def undo(self):
        self.light.off()
        self.light = setUndo(self.light, self.prevBright)
        
# Light Off command
class LightOffCommand(Command):
    def __init__(self, light):
        if isinstance(light, Light):
            self.light = light
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.light.getState()
        self.light.off()
        
    def undo(self):
        self.light.on()
        self.light = setUndo(self.light, self.prevBright)
        
